### PR TITLE
[WIP]New modules to list AWS API Gateway.

### PR DESCRIPTION
**Pull request by cahlchang**
##### SUMMARY

Added list module for AWS API Gateway.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
aws_api_gateway_facts

##### ANSIBLE VERSION
```
(myenv) $ ansible --version
ansible 2.9.0.dev0 config file = None configured module search path =
['/Users/myuser/.ansible/plugins/modules',
'/usr/share/ansible/plugins/modules'] ansible python module location =
/Users/myuser/git/ansible/lib/ansible executable location = ./bin/ansible
python version = 3.7.3 (default, Mar 27 2019, 09:23:15) [Clang 10.0.1
(clang-1001.0.46.3)]
```

##### ADDITIONAL INFORMATION

```paste below

- hosts: localhost
  tasks:
    - aws_api_gateway_facts:
        profile: default
        api_id: 'myapigatewayid'
      register: facts1

    - aws_api_gateway_facts:
        profile: default
      register: facts2

    - debug:
        msg: "{{ facts1 }}"

    - debug:
        msg: "{{ facts2 }}"
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/aws_api_gateway_facts.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      sample: 2018-12-08T13:30:58Z
+
+'''
+
+import traceback
+
+try:
+    import boto3
+    from botocore.exceptions import ClientError, BotoCoreError
+    HAS_BOTO3 = True
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_api_gateway_facts.py:77]: HAS_BOTO3 is not needed with AnsibleAWSModule

File: lib/ansible/modules/cloud/amazon/aws_api_gateway_facts.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+except ImportError:
+    HAS_BOTO3 = False
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import (boto3_conn, ec2_argument_spec, get_aws_connection_info, camel_dict_to_snake_dict)
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_api_gateway_facts.py:81]: boto3_conn, ec2_argument_spec, get_aws_connection_info
are not needed with AnsibleAWSModule

File: lib/ansible/modules/cloud/amazon/aws_api_gateway_facts.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+    try:
+        if api_id:
+            result = client.get_rest_api(restApiId=api_id)
+            del result['ResponseMetadata']
+            api_result = {'items': [result]}
+        else:
+            api_result = client.get_rest_apis()
+    except (ClientError, BotoCoreError) as err:
+        module.fail_json(msg=err.message, exception=traceback.format_exc(), **camel_dict_to_snake_dict(err.response))
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_api_gateway_facts.py:96]: use `module.fail_json_aws(msg=err.message)`

File: lib/ansible/modules/cloud/amazon/aws_api_gateway_facts.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    module = AnsibleAWSModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+    )
+
+    api_id = module.params.get('api_id')
+
+    region, ec2_url, aws_connect_params = get_aws_connection_info(module, boto3=True)
+
+    if region:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_api_gateway_facts.py:121]: just use `client = module.client('apigateway')`

File: lib/ansible/modules/cloud/amazon/aws_api_gateway_facts.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+    exit_args = {"changed": False}
+
+    try:
+        if api_id:
+            result = client.get_rest_api(restApiId=api_id)
+            del result['ResponseMetadata']
+            api_result = {'items': [result]}
+        else:
+            api_result = client.get_rest_apis()
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_api_gateway_facts.py:94]: You'll need to handle pagination if there are a lot of APIs. Should probably handle rate limiting here too

https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#api-throttling-rate-limiting-and-pagination

File: lib/ansible/modules/cloud/amazon/aws_api_gateway_facts.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  aws_api_gateway_facts:
+    api_id: xxxxxxxxxx
+'''
+
+RETURN = '''
+apis:
+  description: a list of api-gateway
+  returned: always
+  type: complex
+  contains:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_api_gateway_facts.py:48]: This does not seem to be at all comprehensive compared to what is in https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/apigateway.html#APIGateway.Client.get_rest_apis

You can generate the RETURN list using hacking/return_skeleton_generator.py

**Review by willthames — CHANGES_REQUESTED** [review]

**Comment by cahlchang** [comment]
Thank you for the review.
I will fix the contents and submit it in the near future.

**Pull request closed without merging**